controller/object_store_controller.go

```diff
@@ -0,0 +1,1139 @@
+		if err != nil {
+			store.Status.State = longhorn.ObjectStoreStateError
+			return nil, store, errors.Wrap(err, "failed to create longhorn volume")
+		} else if store.Status.State != longhorn.ObjectStoreStateStarting {
+			store.Status.State = longhorn.ObjectStoreStateStarting
+		}
+		return vol, store, nil
+	}
+	return nil, store, err
+}
+
+func (osc *ObjectStoreController) checkVolume(vol *longhorn.Volume) error {
+	return nil
+}
+
+func (osc *ObjectStoreController) getOrCreatePV(
+	store *longhorn.ObjectStore,
+	volume *longhorn.Volume,
+) (*corev1.PersistentVolume, *longhorn.ObjectStore, error) {
+	pv, err := osc.ds.GetPersistentVolume(genPVName(store))
+	if err == nil {
+		return pv, store, nil
+	} else if datastore.ErrorIsNotFound(err) {
+		pv, err = osc.createPV(store, volume)
+		if err != nil {
+			store.Status.State = longhorn.ObjectStoreStateError
+			return nil, store, errors.Wrap(err, "failed to create persistent volume")
+		} else if store.Status.State != longhorn.ObjectStoreStateStarting {
+			store.Status.State = longhorn.ObjectStoreStateStarting
+		}
+		return pv, store, nil
+	}
+	return nil, store, err
+}
+
+func (osc *ObjectStoreController) checkPV(pv *corev1.PersistentVolume) error {
+	return nil
+}
+
+func (osc *ObjectStoreController) getOrCreateDeployment(store *longhorn.ObjectStore) (*appsv1.Deployment, *longhorn.ObjectStore, error) {
+	dpl, err := osc.ds.GetDeployment(store.Name)
+	if err == nil {
+		return dpl, store, nil
+	} else if datastore.ErrorIsNotFound(err) {
+		dpl, err = osc.createDeployment(store)
+		if err != nil {
+			store.Status.State = longhorn.ObjectStoreStateError
+			return nil, store, errors.Wrap(err, "failed to create deployment")
+		} else if store.Status.State != longhorn.ObjectStoreStateStarting {
+			store.Status.State = longhorn.ObjectStoreStateStarting
+		}
+		return dpl, store, nil
+	}
+	return nil, store, err
+}
+
+func (osc *ObjectStoreController) checkDeployment(deployment *appsv1.Deployment) error {
+	if *deployment.Spec.Replicas != 1 {
+		deployment.Spec.Replicas = int32Ptr(1)
+		_, err := osc.ds.UpdateDeployment(deployment)
+		return err
+	} else if deployment.Status.UnavailableReplicas > 0 {
+		return ErrObjectStoreDeploymentNotReady
+	}
+	return nil
+}
+
+func (osc *ObjectStoreController) getOrCreateService(store *longhorn.ObjectStore) (*corev1.Service, *longhorn.ObjectStore, error) {
+	svc, err := osc.ds.GetService(osc.namespace, store.Name)
+	if err == nil {
+		return svc, store, nil
+	} else if datastore.ErrorIsNotFound(err) {
+		svc, err = osc.createService(store)
+		if err != nil {
+			store.Status.State = longhorn.ObjectStoreStateError
+			return nil, store, errors.Wrap(err, "failed to create service")
+		} else if store.Status.State != longhorn.ObjectStoreStateStarting {
+			store.Status.State = longhorn.ObjectStoreStateStarting
+		}
+		return svc, store, nil
+	}
+	return nil, store, err
+}
+
+func (osc *ObjectStoreController) checkService(svc *corev1.Service) error {
```

Comment:
Can we do a healthcheck request to the service endpoint to make sure that the S3 endpoint is actually up and running here?

Reply:
Yes, that is the intention of this stub function. I've yet to put a healthcheck implementation there tough ;-)